pkg/proctree/process.go

```diff
@@ -0,0 +1,383 @@
+	userId int,
+	containerId string,
+) {
+	p.setUserId(userId)
+	p.setContainerId(containerId)
+	p.setNsId(nsPid)
+}
+
+// addThreadBasic add the thread to the process node if it does not exist.
+// The function also tries to synchronize the thread exit time with the process if filled after
+// process exit.
+// This function *does not* add the thread to the process tree, so it should be added afterward.
+func (p *processNode) addThreadBasic(tid int) (*threadNode, error) {
+	t, exist := p.getThread(tid)
+	if exist {
+		return t, nil
+	}
+	var err error
+	t, err = newThreadNode(tid)
+	if err != nil {
+		return nil, err
+	}
+	// Update thread exit time to match process if process exited
+	t.setExitTime(p.getExitTime())
+	t.connectToProcess(p)
+	p.connectToThread(t)
+	return t, nil
+}
+
+// isAlive return if the process is alive at the given moment, according to existing information
+// of the node.
+func (p *processNode) isAlive(checkTime types.Timestamp) bool {
+	exitTime := p.getExitTime()
+	if exitTime == 0 {
+		return true
+	}
+	if checkTime >= exitTime {
+		return false
+	}
+	forkTime := p.getForkTime()
+	return checkTime >= forkTime
+}
+
+// setExitTime sets the process's exit time
+func (p *processNode) setExitTime(exitTime types.Timestamp) {
+	p.exitTime = exitTime
+}
+
+// setDefaultExitTime sets the process's exit time if it's not initialized
+func (p *processNode) setDefaultExitTime(exitTime types.Timestamp) {
+	if p.exitTime == 0 {
+		p.exitTime = exitTime
+	}
+}
+
+// exited return if the process exit was received
+func (p *processNode) exited() bool {
+	return p.getExitTime() != 0
+}
+
+// getExitTime return the process's exit time
+func (p *processNode) getExitTime() types.Timestamp {
+	return p.exitTime
+}
+
+// setForkTime sets the process's fork time
+func (p *processNode) setForkTime(forkTime types.Timestamp) {
+	p.forkTime = forkTime
+}
+
+// setDefaultForkTime sets the process's fork time if it's not initialized
+func (p *processNode) setDefaultForkTime(forkTime types.Timestamp) {
+	if p.forkTime == 0 {
+		p.forkTime = forkTime
+	}
+}
+
+// getForkTime return the process's fork time
+func (p *processNode) getForkTime() types.Timestamp {
+	return p.forkTime
+}
+
+// fillExecInfo add execution information to the process from raw format
+func (p *processNode) fillExecInfo(
+	binary fileInfo,
+	cmd []string,
+	execTime types.Timestamp,
+) {
+	p.setExecInfo(
+		execTime, procExecInfo{
+			Cmd:             cmd,
+			ExecutionBinary: binary,
+		},
+	)
+}
+
+// setExecInfo add execution information to the process
+func (p *processNode) setExecInfo(execTime types.Timestamp, info procExecInfo) {
+	execState := types.State[procExecInfo]{
+		StartTime: execTime,
+		Val:       info,
+	}
+	p.execInfo.AddState(execState)
+}
+
+// setDefaultExecInfo change the execution information assumed for the process before its first
+// execution received.
+func (p *processNode) setDefaultExecInfo(info procExecInfo) {
+	p.execInfo.ChangeDefault(info)
```

Comment:
Like here, for example. You have the internal to execInfo mutex, protecting the execInfo internal fields, but you still have a Load() operation for the pointer when you de-reference `p.execInfo`. So the correct thread-safety operation here would either use the mutex or use an atomic Load for a `*Struct{}` (*procExecInfo in this case). Just like an int64 should have an atomic Load()/Store(), etc.

Reply:
Then we can argue on fields that are changed runtime and the fields that aren't changed runtime. If that is the case (fields that are never changed runtime), we should make it very explicity they don't need to be atomic because they're **immutable**.

Reply:
Again, the problem you point here is not relevant because this is not a thread-safe operation.
I think that maybe you are confused because the `TimeSeries` object is thread-safe, and I use it. I implemented it that way so it will be safe to use in all places, not specifically in the current process tree nodes which are not thread-safe.
The face I use a thread-safe object in my nodes doesn't mean the nodes are threa-safe.
I will add a documentation for it on the nodes so it will be clear that they are not thread-safe.